### PR TITLE
[Doc] Convert `<Datagrid>` documentation to TS

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: docs/Datagrid.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ The `Datagrid` component renders a list of records as a table. It is usually use
 
 `<Datagrid>` renders as many columns as it receives `<Field>` children. It uses the field `label` as column header (or, for fields with no `label`, the field `source`).
 
-```jsx
+```tsx
 // in src/posts.js
```
Comment (slax57) [inline docs/Datagrid.md:17]:
```suggestion
// in src/posts.tsx
```

File: docs/Datagrid.md [hunk 2]
```diff
@@ -65,44 +65,71 @@ By default, `<Datagrid>` renders its body using `<DatagridBody>`, an internal re
 
 For instance, the `<Datagrid isRowSelectable>` prop allows to disable the selection checkbox for some records. To *hide* checkboxes instead of disabling them, you can override `<DatagridRow>` and `<DatagridBody>` as follows:
 
-```jsx
+```tsx
 // in src/PostList.js
```
Comment (slax57) [inline docs/Datagrid.md:69]:
```suggestion
// in src/PostList.tsx
```

File: docs/Datagrid.md [hunk 3]
```diff
@@ -157,7 +184,7 @@ Bulk action button components can use the [`useListContext`](./useListContext.md
 
 Here is an example of custom bulk action button, which sets the `views` property of all posts to `0`:
 
-```jsx
+```tsx
 // in ./ResetViewsButton.js
```
Comment (slax57) [inline docs/Datagrid.md:188]:
```suggestion
// in ./ResetViewsButton.tsx
```

File: docs/Datagrid.md [hunk 4]
```diff
@@ -187,7 +214,7 @@ const ResetViewsButton = () => (
 
 But let's say you need a customized bulkAction button. Here is an example leveraging the `useUpdateMany` hook, which sets the `views` property of all posts to `0`:
 
-```jsx
+```tsx
 // in ./CustomResetViewsButton.js
```
Comment (slax57) [inline docs/Datagrid.md:218]:
```suggestion
// in ./CustomResetViewsButton.tsx
```

File: docs/Datagrid.md [hunk 5]
```diff
@@ -233,7 +260,7 @@ export default CustomResetViewsButton;
 
 But most of the time, bulk actions are mini-applications with a standalone user interface (in a Dialog). Here is the same `<CustomResetViewsAction>` implemented behind a confirmation dialog:
 
-```jsx
+```tsx
 // in ./CustomResetViewsButton.js
```
Comment (slax57) [inline docs/Datagrid.md:264]:
```suggestion
// in ./CustomResetViewsButton.tsx
```

File: docs/Datagrid.md [hunk 6]
```diff
@@ -352,7 +379,7 @@ What's a Field component? Simply a component that reads the record (via `useReco
 
 You can even create your own field components.
 
-```jsx
+```tsx
 // in src/users.js
```
Comment (slax57) [inline docs/Datagrid.md:383]:
```suggestion
// in src/users.tsx
```

File: docs/Datagrid.md [hunk 7]
```diff
@@ -512,28 +539,29 @@ By default, `<Datagrid>` renders the table head using `<DatagridHeader>`, an int
...
+import { Children, isValidElement } from "react";
+import { DatagridHeaderProps, FieldProps, List, Datagrid } from "react-admin";
+
+const DatagridHeader = ({ children }: DatagridHeaderProps) => (
+	<TableHead>
+		<TableRow>
+			<TableCell></TableCell>{" "}
+			{/* empty cell to account for the select row checkbox in the body */}
+			{Children.map(children, (child) =>
+				isValidElement<FieldProps>(child) ? (
```
Comment (adguernier) [inline docs/Datagrid.md:553]: Why not using `React.isValidElement` as other snippets ?

File: docs/Datagrid.md [hunk 8]
```diff
@@ -1176,7 +1206,7 @@ The column headers are buttons allowing users to change the list sort field and
 It is possible to disable sorting for a specific `<Field>` by passing a `sortable` property set to `false`:
 
 {% raw %}
-```jsx
+```tsx
 // in src/posts.js
```
Comment (slax57) [inline docs/Datagrid.md:1210]:
```suggestion
// in src/posts.tsx
```

File: docs/Datagrid.md [hunk 9]
```diff
@@ -1197,7 +1227,7 @@ export const PostList = () => (
 By default, a column is sorted by the `source` property. To define another attribute to sort by, set it via the `<Field sortBy>` property:
 
 {% raw %}
-```jsx
+```tsx
 // in src/posts.js
```
Comment (slax57) [inline docs/Datagrid.md:1231]:
```suggestion
// in src/posts.tsx
```

File: docs/Datagrid.md [hunk 10]
```diff
@@ -1223,7 +1253,7 @@ export const PostList = () => (
 
 By default, when the user clicks on a column header, the list becomes sorted in the ascending order. You change this behavior by setting the `sortByOrder` prop to `"DESC"` in a `<Datagrid>` `<Field>`:
 
-```jsx
+```tsx
 // in src/posts.js
```
Comment (slax57) [inline docs/Datagrid.md:1257]:
```suggestion
// in src/posts.tsx
```

**Review by slax57 — CHANGES_REQUESTED** [review]

File: docs/Datagrid.md [hunk 11]
```diff
@@ -512,28 +539,29 @@ By default, `<Datagrid>` renders the table head using `<DatagridHeader>`, an int
...
-                <TableCell key={child.props.source}>
-                    {child.props.source}
-                </TableCell>
-            ))}
-        </TableRow>
-    </TableHead>
+```tsx
+import { TableHead, TableRow, TableCell } from "@mui/material";
+import { Children, isValidElement } from "react";
+import { DatagridHeaderProps, FieldProps, List, Datagrid } from "react-admin";
```
Comment (adguernier) [inline docs/Datagrid.md:545]:
```suggestion
import { DatagridHeaderProps, FieldProps, List, Datagrid, DatagridHeader } from "react-admin";
```

File: docs/Datagrid.md [hunk 12]
```diff
@@ -512,27 +539,28 @@ By default, `<Datagrid>` renders the table head using `<DatagridHeader>`, an int
...
 
-```jsx
-import { TableHead, TableRow, TableCell } from '@mui/material';
+```tsx
+import * as React from "react";
+import { TableHead, TableRow, TableCell } from "@mui/material";
+import { DatagridHeaderProps, FieldProps, List, Datagrid, DatagridHeader } from "react-admin";
 
-const DatagridHeader = ({ children }) => (
+const DatagridHeader = ({ children }: DatagridHeaderProps) => (
```
Comment (slax57) [inline docs/Datagrid.md:547]: In this snippet you are both importing `DatagridHeader` from `"react-admin"` and declaring `const DatagridHeader`. This cannot work. You should remove the import I believe.

**Review by slax57 — CHANGES_REQUESTED** [review]

File: docs/Datagrid.md [hunk 13]
```diff
@@ -512,27 +539,28 @@ By default, `<Datagrid>` renders the table head using `<DatagridHeader>`, an int
 
 For instance, here is a simple datagrid header that displays column names with no sort and no "select all" button:
 
-```jsx
-import { TableHead, TableRow, TableCell } from '@mui/material';
+```tsx
+import * as React from "react";
+import { TableHead, TableRow, TableCell } from "@mui/material";
+import { DatagridHeaderProps, FieldProps, List, Datagrid, DatagridHeader } from "react-admin";
```
Comment (djhi) [inline docs/Datagrid.md:545]:
```suggestion
import { DatagridHeaderProps, FieldProps, List, Datagrid } from "react-admin";
```

**Review by slax57 — APPROVED** [review]

**Pull request merged**